### DISCUSSION TITLE
Is it actually possible to implement alarms for ios?

### DISCUSSION THREAD

**Discussion by LesYampolskyi** [gdelataillade/alarm#36] [Q&A] [answered]
I am not sure, if it is actually possible to implement an alarm for ios for a long time background mode. I tested it for 15 minutes, and it works. But what about the longer term? Several hours or days.

There is a warning in xcode's console:

>Background Task 3 ("Flutter debug task"), was created over 30 seconds ago. In applications running in the background, this creates a risk of termination. Remember to call UIApplication.endBackgroundTask(_:) for your task in a timely manner to avoid this.

<code>DispatchQueue.main.asyncAfter</code> will not trigger after 30 seconds. And maybe after some period even <code>self.audioPlayers[id]!.play(atTime: time)</code> neither.

It needs to be tested on Testflight or Firebase console. (build in profile or release modes to enable launching from the home screen.)

And there is a similar question in [stackoverflow](https://stackoverflow.com/questions/43538553/is-it-possible-to-build-an-alarm-clock-such-as-the-built-in-app-from-apple) about alarms. The topic is quite old (6 years), but there is an opinion that it is impossible.

>Unfortunately this is unavailable for developers. You can check all of alarm clock apps in AppStore, all of them are using the standard local notifications.

And another one about App ["Alarmy"](https://stackoverflow.com/questions/22823126/app-alarmy-is-able-to-play-itunes-song-from-background-state-how)

>They use the "NoSound.mp3" file in their bundle to play silence, while in the background. 

This may consume additional battery. And will this hack cause problems with publishing in the App Store?

**Comment by LesYampolskyi** [comment, marked as answer]
Sorry. I inadvertently read the iOS section of the README.md and didn't notice 
>Due to the suspension of asynchronous native code when the app is in the background, we listen for app state changes and check if the player is playing when the app returns to the foreground.

It might be better to explicitly state that the alarm would not work on ios in the background. 
You could delete this discussion.

**Comment by gdelataillade** [comment]
I did a lot of research and you're right, there is always the risk that iOS decides to kill the native background processes to free up some RAM if needed. Unfortunately, I couldn't find a way that would be 100% reliable. This is how the OS is designed so the only things we can do is find some hacks.
